### PR TITLE
Automatically enable depends of mods when enabling the mods in the world config menu

### PR DESCRIPTION
**Pull request by HybridDog**
the hard dependencies are enabled, the optional ones need to be enabled manually

### CODE REVIEW

**Comment by HybridDog** [comment]
rebased

File: builtin/mainmenu/dlg_config_world.lua [hunk 1]
```diff
@@ -91,30 +80,94 @@ local function get_formspec(data)
...
+		if depends then
+			for _,name in pairs(depends) do
+				if not enabled_mods[name] then
+					to_enable[n] = name
+					n = n+1
+				end
+			end
+		end
+	end
+	while next(to_enable) do
```
Comment (HybridDog) [inline builtin/mainmenu/dlg_config_world.lua]: l need to update this loop

**Comment by HybridDog** [comment]
Shall l squash the commits?

**Comment by HybridDog** [comment]
@sfan5 rebased

File: builtin/mainmenu/dlg_config_world.lua [hunk 2]
```diff
@@ -22,30 +22,14 @@ end
...
-	end
-
-	if mod == nil then
-		mod = {name=""}
-	end
-	
-	retval = retval ..
+		"label[1.75,0;" .. data.worldspec.name .. "]"..
+		"checkbox[1,6;cb_hide_gamemods;" .. fgettext("Hide Game") .. ";" .. tostring(data.hide_gamemods) .. "]" ..
+		"checkbox[6,6;cb_hide_mpcontent;" .. fgettext("Hide mp content") .. ";" .. tostring(data.hide_modpackcontents) .. "]" ..
```
Comment (SmallJoker) [inline builtin/mainmenu/dlg_config_world.lua]: These lines are too long (> 80 characters)

File: builtin/mainmenu/dlg_config_world.lua [hunk 3]
```diff
@@ -54,137 +38,187 @@ local function get_formspec(data)
...
-			if mod.enabled then
-				retval = retval .. "checkbox[5.5,-0.125;cb_mod_enable;" .. fgettext("enabled") .. ";true]"
 			else
-				retval = retval .. "checkbox[5.5,-0.125;cb_mod_enable;" .. fgettext("enabled") .. ";false]"
+				retval = retval .. "button[5.5,0.125;2.5,0.5;btn_mp_enable;" .. fgettext("Enable MP") .. "]"
 			end
+		elseif mod.enabled then
+			retval = retval .. "checkbox[5.5,-0.125;cb_mod_enable;" .. fgettext("enabled") .. ";true]"
+		else
+			retval = retval .. "checkbox[5.5,-0.125;cb_mod_enable;" .. fgettext("enabled") .. ";false]"
```
Comment (SmallJoker) [inline builtin/mainmenu/dlg_config_world.lua]: You simplified "Hide Game" and "Hide mp content" but not this enable checkbox - why?

File: builtin/mainmenu/dlg_config_world.lua [hunk 4]
```diff
@@ -248,62 +280,59 @@ function create_configure_world_dlg(worldidx)
...
-				end, --filter
-				{ worldpath= dlg.data.worldspec.path,
-				  gameid = dlg.data.worldspec.gameid }
-			)
+				end
+			end,
+			function(element,criteria)
+				if criteria.hide_game and
+					element.typ == "game_mod" then
+						return false
```
Comment (SmallJoker) [inline builtin/mainmenu/dlg_config_world.lua]: Indents should be the other way, like this:
```lua
if criteria.hide_game
		and element.typ == "game_mod" then
	return false
end
```

File: builtin/mainmenu/dlg_config_world.lua [hunk 5]
```diff
@@ -248,62 +280,59 @@ function create_configure_world_dlg(worldidx)
...
+				end
+
+				if criteria.hide_modpackcontents and
+					element.modpack ~= nil then
+						return false
+					end
```
Comment (SmallJoker) [inline builtin/mainmenu/dlg_config_world.lua]: Same issue as above

File: builtin/mainmenu/dlg_config_world.lua (hunk 2 above)
Comment (HybridDog) [inline builtin/mainmenu/dlg_config_world.lua]: fixed

File: builtin/mainmenu/dlg_config_world.lua (hunk 3 above)
Comment (HybridDog) [inline builtin/mainmenu/dlg_config_world.lua]: l didn't notice it because l continued at the end after unwrapping the inner if statement.

File: builtin/mainmenu/dlg_config_world.lua (hunk 4 above)
Comment (HybridDog) [inline builtin/mainmenu/dlg_config_world.lua]: fixed, below too

**Comment by HybridDog** [comment]
rebased, the soft depends are enabled, the hard ones aren't

File: builtin/mainmenu/dlg_config_world.lua [hunk 6]
```diff
@@ -24,38 +24,72 @@ local function modname_valid(name)
...
+			info = fgettext("No modpack description provided.")
+		end
+		retval = retval ..
+			"label[0,0.7;" .. info .. "]"
+	else
+		retval = retval ..
+			"label[0,0.7;" .. fgettext("Mod:") .. "]" ..
+			"label[0.75,0.7;" .. mod.name .. "]"
+
+		if hard_deps == "" then
```
Comment (rubenwardy) [inline builtin/mainmenu/dlg_config_world.lua]: I think it's fine to show an empty textlist

File: builtin/mainmenu/dlg_config_world.lua [hunk 7]
```diff
@@ -69,59 +103,135 @@ local function get_formspec(data)
...
-	retval = retval .. ";" .. data.selected_mod .."]"
+		"table[5.5,0.75;5.75,6;world_config_modlist;" ..
+		modmgr.render_modlist(data.list) ..
+		";" .. data.selected_mod .."]"
+end
 
-	return retval
+-- makes a list of mod ids indexed by their names
+local mod_ids
+local function get_mod_ids(list)
```
Comment (rubenwardy) [inline builtin/mainmenu/dlg_config_world.lua]: `update_mod_ids`, as this doesn't return a value

File: builtin/mainmenu/dlg_config_world.lua [hunk 8]
```diff
@@ -69,59 +103,135 @@ local function get_formspec(data)
...
+		return
+	end
 
+	-- enable mod's depends
+
+	if not mod_ids then
+		get_mod_ids(list)
+	end
+
+	local to_enable,n = {},1
```
Comment (rubenwardy) [inline builtin/mainmenu/dlg_config_world.lua]: I don't like this style

File: builtin/mainmenu/dlg_config_world.lua [hunk 9]
```diff
@@ -69,59 +103,135 @@ local function get_formspec(data)
...
+			--~ for i = 1,#optinal_depends do
+				--~ local name = optinal_depends[i]
+				--~ if not enabled_mods[name] then
+					--~ to_enable[n] = name
+					--~ n = n+1
+				--~ end
+			--~ end
+		end
+	end
+	if n == 1 then
```
Comment (rubenwardy) [inline builtin/mainmenu/dlg_config_world.lua]: this would be cleaner if done recursively

File: builtin/mainmenu/dlg_config_world.lua [hunk 10]
```diff
@@ -69,59 +103,135 @@ local function get_formspec(data)
...
 
+	-- enable mod's depends
+
+	if not mod_ids then
+		get_mod_ids(list)
+	end
+
+	local to_enable,n = {},1
+	for name in pairs(enabled_mods) do
+		--~ local depends, optinal_depends = modmgr.get_dependencies(
```
Comment (rubenwardy) [inline builtin/mainmenu/dlg_config_world.lua]: old code

File: builtin/mainmenu/dlg_config_world.lua [hunk 11]
```diff
@@ -69,59 +103,135 @@ local function get_formspec(data)
...
+		-- no mod found
+		return
+	end
+	n = n-1
+	while n ~= 0 do
+		local name = to_enable[n]
+		n = n-1
+
+		local id = mod_ids[name]
+		if id then
```
Comment (rubenwardy) [inline builtin/mainmenu/dlg_config_world.lua]: I'd remove this then do:

```lua
local mod = list[mod_ids[name]]
if mod then
```

`list[nil] == nil`

File: builtin/mainmenu/dlg_config_world.lua [hunk 12]
```diff
@@ -159,24 +267,24 @@ local function handle_buttons(this, fields)
...
 					mod.typ ~= "game_mod" then
 				if modname_valid(mod.name) then
 					worldfile:set("load_mod_"..mod.name, tostring(mod.enabled))
-				else
-					if mod.enabled then
-						gamedata.errormessage = fgettext_ne("Failed to enable mod \"$1\" as it contains disallowed characters. Only chararacters [a-z0-9_] are allowed.", mod.name)
-					end
+				elseif mod.enabled then
+					gamedata.errormessage = fgettext_ne("Failed to enable " ..
+						'mod "$1" as it contains disallowed characters.' ..
```
Comment (rubenwardy) [inline builtin/mainmenu/dlg_config_world.lua]: `"`

File: builtin/mainmenu/init.lua [hunk 13]
```diff
@@ -16,9 +16,9 @@
 --51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 
 mt_color_grey  = "#AAAAAA"
-mt_color_blue  = "#0000DD"
-mt_color_green = "#00DD00"
-mt_color_dark_green = "#003300"
+mt_color_blue  = "#6389FF"
```
Comment (rubenwardy) [inline builtin/mainmenu/init.lua]: unrelated change, should be a separate commit / PR

File: builtin/mainmenu/modmgr.lua [hunk 14]
```diff
@@ -279,34 +276,51 @@ function modmgr.render_modlist(render_list)
...
-					table.insert(hard_dependencies, dependency)
-				end
-				dependency = dependencyfile:read()
-			end
-			dependencyfile:close()
+function modmgr.get_modpack_description(modfolder)
+	assert(modfolder, "missing modpack folder")
+	local modpackfile = io.open(modfolder .. DIR_DELIM .. "modpack.txt", "r")
+	assert(modpackfile, "missing modpack.txt")
+	local content = modpackfile:read("*all")
```
Comment (rubenwardy) [inline builtin/mainmenu/modmgr.lua]: modpack.txt does not usually contain a description. Use description.txt

File: builtin/mainmenu/modmgr.lua [hunk 15]
```diff
@@ -279,34 +276,51 @@ function modmgr.render_modlist(render_list)
...
+	modpackfile:close()
+	return minetest.formspec_escape(content)
+end
+
+--------------------------------------------------------------------------------
+function modmgr.get_dependencies(modfolder, get_table)
+	--~ assert(modfolder, "modmgr.get_dependencies: missing modfolder")
+	if not modfolder then
+		return "modmgr.get_dependencies: missing modfolder",
```
Comment (rubenwardy) [inline builtin/mainmenu/modmgr.lua]: not good to return like this

File: builtin/mainmenu/dlg_config_world.lua (hunk 6 above)
Comment (HybridDog) [inline builtin/mainmenu/dlg_config_world.lua]: I did this change because some mods have lots of optional but no hard dependency (and the other way round), so the textlist got a scrollbar and became inconvenient.

File: builtin/mainmenu/dlg_config_world.lua (hunk 7 above)
Comment (HybridDog) [inline builtin/mainmenu/dlg_config_world.lua]: ok..

File: builtin/mainmenu/dlg_config_world.lua (hunk 12 above)
Comment (HybridDog) [inline builtin/mainmenu/dlg_config_world.lua]: ok

File: builtin/mainmenu/dlg_config_world.lua (hunk 6 above)
Comment (HybridDog) [inline builtin/mainmenu/dlg_config_world.lua]: 2017-02-11-122533_1920x1080_scrot
2017-02-11-121451_1920x1080_scrot

**Comment by HybridDog** [comment]
> this would be cleaner if done recursively

it would require an additional function

> not good to return like this

Do l need additional indentation there?

**Comment by HybridDog** [comment]
> this would be cleaner if done recursively

it would require an additional function

> not good to return like this

Do l need additional indentation there?

**Comment by rubenwardy** [comment]
It should return nil if problem getting dependencies, returning an error message like that is odd imo

**Comment by rubenwardy** [comment]
It's fine to add new functions. The most important thing is that it's readable and maintainable

**Comment by HybridDog** [comment]
>It should return nil if problem getting dependencies, returning an error message like that is odd imo

I made it abort when such a problem happens before. But l noticed a bug: selecting game mods doesn't work right. If you select the last one (l think), it aborts and every time you click on configure minetest aborts, which is worse than having a strange return value until the bug is fixed.

File: builtin/mainmenu/dlg_config_world.lua (hunk 6 above)
Comment (DS-Minetest) [inline builtin/mainmenu/dlg_config_world.lua]: I like the "No optional dependencies" but then for consistence it should be the same with hard dependencies.

**Comment by DS-Minetest** [comment]
:+1: 
This PR seems to be great. Many people would love the new features.

**Comment by rubenwardy** [comment]
are you still willing to work on this?

**Comment by HybridDog** [comment]
What work do you mean?

**Comment by octacian** [comment]
I like the concept, but I would also prefer if actual enabling is done in the `modmgr`. In other words, I think the implementation could be improved. On the negative side, this is gonna make my life incredibly fun to factor in the changes of this PR to my complete rewrite of `modmgr` xD

On sidenote though, I won't be able to finish that PR for a least another month, so wouldn't bother me if this PR goes ahead.

File: builtin/mainmenu/modmgr.lua (hunk 15 above)
Comment (HybridDog) [inline builtin/mainmenu/modmgr.lua]: Is that subgame mod selection fixed yet?

**Comment by paramat** [comment]
I suspect there are times when, for certain testing and development reasons, you don't want this automatically done, so perhaps needs to have a setting.

**Comment by rubenwardy** [comment]
I'd rather the code was cleaned up first before adding this feature, as it's not a small change and makes refactoring a little harder. Shame that @octacian is busy :(

**Comment by octacian** [comment]
> I suspect there are times when, for certain testing and development reasons, you don't want this automatically done, so perhaps needs to have a setting.

In general, while developer ease is important, Minetest should really be optimized for the player more than the developer. Although, I'm not against tools to make things easier for the developer (i.e. a setting would be okay, but not required IMO).

**Comment by HybridDog** [comment]
> I suspect there are times when, for certain testing and development reasons, you don't want this automatically done, so perhaps needs to have a setting.

Developers usually are aware about how mods are en-/disabled. In that certain case, l doubt it happens frequently, developers can manually edit depends.txt or world.mt.
Also, the dependencies which become enabled automatically are listed on the left side in the configuration formspec, so people see which mods are to be enabled automatically (except the dependencies of dependencies l think).

**Comment by HybridDog** [comment]
> I'd rather the code was cleaned up first before adding this feature, as it's not a small change and makes refactoring a little harder.

Me too, please merge #5351 before this.

**Comment by paramat** [comment]
#5351 is to be merged before this, then this can be rebased on that.

**Comment by HybridDog** [comment]
rebased

**Comment by paramat** [comment]
#5351 is merged and this rebased on that, so @rubenwardy 's request seems done.

**Comment by paramat** [comment]
@HybridDog

File: builtin/mainmenu/pkgmgr.lua [hunk 16]
```diff
@@ -320,33 +324,100 @@ function pkgmgr.is_modpack_entirely_enabled(data, name)
 	return true
 end
 
----------- toggles or en/disables a mod or modpack -----------------------------
+--------------------------------------------------------------------------------
+-- makes a list of mod ids indexed by their names
+local mod_ids
+local function update_mod_ids(list)
+	mod_ids = {}
```
[the remark this answers is not in the document]
Comment (HybridDog) [reply builtin/mainmenu/pkgmgr.lua]: You're right.

**Comment by HybridDog** [comment]
Shall I print the enabled mods to action log so that you can see in the terminal which mods were enabled?

**Comment by HybridDog** [comment]
It now prints the en- or disabled mods to action log so that you can see in the terminal which mods were enabled.

File: builtin/mainmenu/dlg_config_world.lua (hunk 7 above)
Comment (HybridDog) [reply builtin/mainmenu/dlg_config_world.lua]: This function is no longer used.

**Comment by HybridDog** [comment]
@rubenwardy, are you fine with the changes?

File: builtin/mainmenu/pkgmgr.lua [hunk 17]
```diff
@@ -350,35 +347,111 @@ function pkgmgr.is_modpack_entirely_enabled(data, name)
...
+	end
+
+	-- to_enable is used as a DFS stack with sp as stack pointer
+	local to_enable = {}
+	local sp = 0
+	for name in pairs(enabled_mods) do
+		local depends = pkgmgr.get_dependencies(list[mod_ids[name]].path)
+		for i = 1, #depends do
+			local name = depends[i]
+			if not enabled_mods[name] then
```
Comment (DS-Minetest) [inline builtin/mainmenu/pkgmgr.lua]: Why not `name, is_enabled` in the for loop and `not is_enabled` here?

File: builtin/mainmenu/pkgmgr.lua [hunk 18]
```diff
@@ -350,35 +347,111 @@ function pkgmgr.is_modpack_entirely_enabled(data, name)
...
+		end
+	end
+
+	-- to_enable is used as a DFS stack with sp as stack pointer
+	local to_enable = {}
+	local sp = 0
+	for name in pairs(enabled_mods) do
+		local depends = pkgmgr.get_dependencies(list[mod_ids[name]].path)
+		for i = 1, #depends do
+			local name = depends[i]
```
Comment (DS-Minetest) [inline builtin/mainmenu/pkgmgr.lua]: I suggest to give this another name than `name` to avoid confusion with the name of the name of the depending mod called `name`.

File: builtin/mainmenu/pkgmgr.lua [hunk 19]
```diff
@@ -350,35 +347,111 @@ function pkgmgr.is_modpack_entirely_enabled(data, name)
...
+			else
+				if mod.enabled == false then
+					mod.enabled = true
+					toggled_mods[#toggled_mods+1] = mod.name
+				end
+				-- push the dependencies of the dependency onto the stack
+				local depends = pkgmgr.get_dependencies(mod.path)
+				for i = 1, #depends do
+					sp = sp+1
+					to_enable[sp] = depends[i]
```
Comment (DS-Minetest) [inline builtin/mainmenu/pkgmgr.lua]: Why is here no `if not enabled_mods[depends[i]] then`?

**Comment by DS-Minetest** [comment]
It would be useful to distinguish between automatically and manually enabled mods and to have something like an autoremove function, but this would be out of scope currently.

File: builtin/mainmenu/pkgmgr.lua (hunk 19 above)
Comment (HybridDog) [reply builtin/mainmenu/pkgmgr.lua]: Neither here, nor in line 419, the condition is required, but it is likely that a dependency is already activated, e.g. many mods depend on default. I've put the test to both places now.

**Comment by HybridDog** [comment]
@rubenwardy, are you fine with the changes?

File: builtin/mainmenu/pkgmgr.lua [hunk 20]
```diff
@@ -350,35 +347,113 @@ function pkgmgr.is_modpack_entirely_enabled(data, name)
...
+					end
+				end
 			end
-			list[i].enabled = toset
 		end
 	end
+
+	-- Log the list of enabled mods
+	table.sort(toggled_mods)
+	minetest.log("action", "Following mods were enabled: " ..
```
Comment (ClobberXD) [inline builtin/mainmenu/pkgmgr.lua]: I'd personally prefer this to be logged at the `info` level.

Comment (HybridDog) [reply builtin/mainmenu/pkgmgr.lua]: I've added this to action because I think that you should easily see which mods are en- or disabled.

Comment (ClobberXD) [reply builtin/mainmenu/pkgmgr.lua]: Good point, but I think that's only relevant while testing this PR, to see if things work as they should. Otherwise, one wouldn't need to look at the logs to know which mods are enabled - they could just see `mod.conf`. :)

Comment (HybridDog) [reply builtin/mainmenu/pkgmgr.lua]: Now it's logged to info.

Comment (ClobberXD) [reply builtin/mainmenu/pkgmgr.lua]: Thank you :)

**Comment by LoneWolfHT** [comment]
*bump*
This would be a nice addition. What's in the way of it being merged? (Besides the requested changes...)

**Comment by HybridDog** [comment]
> What's in the way of it being merged?

Maybe it's because their time is limited. (see https://github.com/minetest/minetest_game/pull/2449#issuecomment-526264843)

**Comment by Fixer-007** [comment]
What if you have several identical mods in different modpacks?

**Comment by HybridDog** [comment]
> What if you have several identical mods in different modpacks?

Minetest aborts if multiple mods with the same name are enabled.
If you enable a modpack with a mod which is already enabled in another modpack, the mod is enabled nonetheless; this is the same behaviour as before as far as I can see.
If the mod which occurs two times is the dependency of a mod which is activated, only one of the two mods is enabled.

**Comment by sfan5** [comment]
:+1: Tested and works

Another nice feature would be to disable mods that depend on X when you disable X.

**Comment by ClobberXD** [comment]
Tested; works. A mod's dependencies are correctly auto-enabled when enabling it. :+1:

****

I tested by having a total of 4 mods:

- basic_materials
- basic_signs
- pipeworks
- signs_lib

Enabling `pipeworks` automatically enabled its hard dependencies `basic_materials`, and enabling `basic_signs` auto-enabled its hard dependencies `basic_materials` (if it wasn't already enabled) and `signs_lib`
